### DISCUSSION TITLE
py5 examples

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#104] [Ideas]
Hey folks, how do you feel about porting Processing Python mode examples to py5? Is it worth it?

In my master's degree dissertation I tried to classify those examples in a taxonomy I was proposing. Unfortunately I somehow broke the [Omeka](https://omeka.org/) repository I was using to host the classified/reorganized examples, and I couldn't fix it yet.

And then, @tabreturn has made this wonderful site, http://pyde.org, so I was wondering if it would be a good idea to port them to py5....

What do you think?

Also, I have been contributing to Jeremy Douglass' Rosetta Code examples https://github.com/jeremydouglass/rosetta_examples_p5 and making support examples for my classes at https://github.com/villares/material-aulas, which I'm planning to port this semester (here in Brazil the teaching semesters are February to June, and August to December).

**Comment by tabreturn** [comment]
My initial response is: where/how would one browse/peruse this 'archive'?

pyde.org was a rough experiment that I kind of abandoned, but something I've been thinking about revisiting. I always found it frustrating to open-then-run examples via *File > Examples* in the Processing IDE (before I get to see what those sketches produce). pyp5js allows one to run Processing.py examples (with relatively minor tweaking) in a web browser, so I wrote something to perform some rudimentary conversion and populate a Jinja template.

Interestingly, the Processing website overhaul now includes a pyde.org-like page (https://processing.org/examples). I'm not claiming pyde.org inspired this -- it was simply the sensible thing to do. The Generative Design book website is another that presents nicely browse-able examples (http://www.generative-gestaltung.de/2/)

Overhaul, I'm thinking: there's a considerable web design and development aspect to this all (although there's a lot we can leverage in py5generator and py5book)? Of course, could we always start converting sketches and storing them in a dedicated repo so long? I'm very interested in taking a look at your taxonomy, @villares. It might also be a good idea to include dedicated sections/categories for some highly popular processing books -- specifically, py5 ports of examples from: _Nature of Code_; _Generative Design_; _Code as Creative Medium_.

**Reply by villares** [reply]
I'm frustrated that I found that not only the Omeka repo is down, but the TemaTres "controlled vocabulary" PHP tool I have deployed is also broken and I have no idea why, after just 3 years!
You can have a look here: https://github.com/villares/mestrado/tree/master/assets
A listing of categories (B) and the examples, with some comments (C): https://github.com/villares/mestrado/blob/master/assets/tesauro-e-exemplos.pdf

I found your approach to pyde.org brilliant, could it host the py5 examples too, perhaps?

**Reply by hx2A** [reply]
> I always found it frustrating to open-then-run examples via _File > Examples_ in the Processing IDE (before I get to see what those sketches produce)

Yes, it is a bit of a dice roll running those examples. The comments do provide some useful information but it is not as nice as a picture.

**Reply by hx2A** [reply]
> It might also be a good idea to include dedicated sections/categories for some highly popular processing books -- specifically, py5 ports of examples from: _Nature of Code_; _Generative Design_; _Code as Creative Medium_.

I agree. the PDE has an "Examples" tab in the Contribution manager and offers packages of examples. Is there some way for Thonny to do something similar?

**Reply by tabreturn** [reply]
> the PDE has an "Examples" tab in the Contribution manager and offers packages of examples. Is there some way for Thonny to do something similar?

I think this would be fairly easy to accomplish in Thonny. I'll make a note to look at this.

**Comment by Maltiec** [comment]
That's an interesting idea, even if we are not sure about the way of storing this archive. For example this archive could be stored in git just for now. I've already translated many sketches from processing python mode to py5 (while preparing lections for my students),  and also transformed some examples from Nature of Code (JAVA to Python). So if we will come to an agreement of a single storing place - I am in. )

**Reply by villares** [reply]
Could it be here perhaps?  https://github.com/py5coding/py5examples
Maybe a branch, while it is an exploratory work in progress...

**Reply by Maltiec** [reply]
btw, we must use Jupiter or it could be .py files?

**Reply by hx2A** [reply]
The py5examples repo was an attempt to share some example code and was created before I knew about Jupyter book. I was just happy to get the Jupyter mybinder service working and see py5 run over the web. Right now it is a bit of a messy dump of notebooks and one *.py file.

In my opinion, one of the benefits of having code in a github repo is the ability to then run that code through a notebook on mybinder. But that isn't how everyone wants to engage with py5. Having individual *.py files is useful also, but would require people to download the code and create a working environment on their machine.

I'm definitely open to ideas for how the repo can be better organized and be a better service to the growing py5 community.

**Reply by tabreturn** [reply]
Can Jupyter handle multi-file sketches? I'm just thinking about sketches that comprise several .py files.

**Reply by Maltiec** [reply]
there are ways for importing functions from one notebook to another, but it kinda tricky. I think more complex sketches is a case for Jupyter Lab

**Reply by hx2A** [reply]
You can write some *.py files and then import them. I believe you can create markdown links that link to other *.py files. It works, but is not as nice as the multiple tabs for files you get in Thonny or the PDE.

**Comment by villares** [comment]
I think we could have both.

I'll try to upload a first attempt of .py files soon.

**Reply by villares** [reply]
Update: I found something strange... very aggressive lowercase conversion...  #111 
```
pp2i = py5_tools.translators.processingpy2imported
result = pp2i.translate_dir(source_dir, target_dir)   # also with ext='.py' for the secondary tabs
print(result)
```
![image](https://user-images.githubusercontent.com/3694604/183500824-b20b14a3-ad6e-41a5-926b-cba497b2256d.png)

I'll try and investigate before uploading to my fork of `py5examples`. I have yet to copy the `data` folders too :(

Meanwhile, totally unrevised: https://github.com/villares/rosetta_examples_p5/tree/py5coding/examples/py5

**Reply by hx2A** [reply]
You are working very hard with the code translations! You are correct, it should not be converting class names like that. I believe your PR that does not touch names that start with an upper case character will properly address this.

**Reply by villares** [reply]
:smile: 
![image](https://user-images.githubusercontent.com/3694604/183527571-bef1e846-1814-48a6-8e5c-c511ace6abd0.png)

On other news, I have copied all the data folders using `pathlib`,  `shutils` and some luck.  :smile:

**Reply by hx2A** [reply]
> On other news, I have copied all the data folders using `pathlib`, `shutils` and some luck. smile

Awesome, glad you have started using `pathlib`!
